File: 54.spiral-matrix.js
/*
 * @lc app=leetcode id=54 lang=javascript
 *
 * [54] Spiral Matrix
 */

// @lc code=start
/**
 * @param {number[][]} matrix
 * @return {number[]}
 */
var spiralOrder = function(matrix) {
  let [top, left] = [0, 0]
  let [bottom, right] = [matrix.length,
                         matrix[0].length]
  let results = []

  while (top < bottom && left < right) {
    // top left to top right
    for (let col = left; col < right; col++) {
      results.push(matrix[top][col])
    }
    top++
    // right top to down right
    for (let row = top; row < bottom; row++) {
      results.push(matrix[row][right - 1])
    }
    right--
    // check for edge cases
    if (!(left < right && top < bottom)) break
    // right bottom to left bottom
    for (let col = right - 1; col > left - 1; col--) {
      results.push(matrix[bottom - 1][col])
    }
    bottom--
    // left bottom to left top
    for (let row = bottom - 1; row > top - 1; row--) {
      results.push(matrix[row][left])
    }
    left++
  }
  return results
};
// @lc code=end

File: cpp/54.spiral-matrix.cpp
/*
 * @lc app=leetcode id=54 lang=cpp
 *
 * [54] Spiral Matrix
 */

// @lc code=start
class Solution
{
public:
    vector<int> spiralOrder(vector<vector<int>> &matrix)
    {
        int top = 0, left = 0;
        int bottom = matrix.size(), right = matrix[0].size();
        vector<int> results;

        while (top < bottom && left < right)
        {
            // top left to top right
            for (int col = left; col < right; col++)
            {
                results.push_back(matrix[top][col]);
            }
            top++;
            // right top to down right
            for (int row = top; row < bottom; row++)
            {
                results.push_back(matrix[row][right - 1]);
            }
            right--;
            // check for edge cases
            if (!(left < right && top < bottom))
                break;
            // right bottom to left bottom
            for (int col = right - 1; col > left - 1; col--)
            {
                results.push_back(matrix[bottom - 1][col]);
            }
            bottom--;
            // left bottom to left top
            for (int row = bottom - 1; row > top - 1; row--)
            {
                results.push_back(matrix[row][left]);
            }
            left++;
        }
        return results;
    }
};
// @lc code=end
